src/guiapi/include/Rect16.h

```diff
@@ -0,0 +1,279 @@
+#pragma once
+
+#include "guitypes.h"
+#include <array>
+#include <climits>
+
+////////////////////////////////////////////////////////////////////////////////
+/// @enum ShiftDir_t
+/// @brief A strongly typed enum class representing the direction where the
+///     the next rectangle will be created
+enum class ShiftDir_t {
+    Left,
+    Right,
+    Top,
+    Bottom
+};
+
+////////////////////////////////////////////////////////////////////////////////
+///
+/// Class prepesents rectagle in graphics system where x rising to the right
+/// and y rising down.
+///
+/// @details Class also support negative value of coordinates. Such a simple perk
+/// makes the implemetation more robust and useful.
+///
+class Rect16 {
+    point_i16_t top_left_;
+    uint16_t width_;
+    uint16_t height_;
+
+public:
+    ////////////////////////////////////////////////////////////////////////////
+    /// @brief Default constructor
+    /// @details set top left corner to {0,0} with width and heigth 0
+    Rect16();
+
+    ////////////////////////////////////////////////////////////////////////////
+    /// @brief Create rectangle on specific top-left and bottom-right
+    ///        corner
+    /// @param[in] x X coordinate of top-left corner
+    /// @param[in] y Y coordinate of top-left corner
+    /// @param[in] x X coordinate of bottom-right corner
+    /// @param[in] y Y coordinate of bottom-right corner
+    Rect16(int16_t, int16_t, int16_t, int16_t);
+
+    ////////////////////////////////////////////////////////////////////////////
+    /// @brief Create rectangle on specific top-left corner and width
+    ///        and heigth
+    /// @param[in] point Top-left corner
+    /// @param[in] width Width in pixels
+    /// @param[in] heigth Heigth in pixels
+    Rect16(point_i16_t, uint16_t, uint16_t);
+
+    ////////////////////////////////////////////////////////////////////////////
+    /// @brief Copy contructor
+    Rect16(Rect16 const &);
```

Comment:
copy constructor will be autogenerated (simple type)
erase it, or use `Rect16(Rect16 const &) = default ;`